moveit_planners/trajopt/src/trajopt_interface.cpp

```diff
@@ -0,0 +1,452 @@
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Omid Heidari */
+
+#include <moveit/planning_interface/planning_interface.h>
+#include <moveit/robot_state/conversions.h>
+
+#include <moveit_msgs/MotionPlanRequest.h>
+
+#include <trajopt_sco/sco_common.hpp>
+#include <trajopt_sco/optimizers.hpp>
+#include <trajopt_sco/solver_interface.hpp>
+
+#include <trajopt/trajectory_costs.hpp>
+
+#include <ros/ros.h>
+#include <rosparam_shortcuts/rosparam_shortcuts.h>
+
+#include <limits>
+#include <vector>
+#include <Eigen/Geometry>
+#include <unordered_map>
+
+#include "trajopt_interface/trajopt_interface.h"
+#include "trajopt_interface/problem_description.h"
+
+namespace trajopt_interface
+{
+TrajOptInterface::TrajOptInterface(const ros::NodeHandle& nh) : nh_(nh), name_("TrajOptInterface")
+{
+  prob_ = TrajOptProblemPtr(new TrajOptProblem);
+  setDefaultTrajOPtParams();
+
+  // TODO: callbacks should be defined by the user
+  callbacks_.push_back(callBackFunc);
+}
+
+bool TrajOptInterface::solve(const planning_scene::PlanningSceneConstPtr& planning_scene,
+                             const planning_interface::MotionPlanRequest& req,
+                             moveit_msgs::MotionPlanDetailedResponse& res)
+{
+  ROS_INFO(" ======================================= From trajopt_interface, solve is called");
+  setTrajOptParams(params_);
+
+  if (!planning_scene)
+  {
+    ROS_ERROR_STREAM_NAMED("trajopt_planner", "No planning scene initialized.");
+    res.error_code.val = moveit_msgs::MoveItErrorCodes::FAILURE;
+    return false;
+  }
+
+  if (req.start_state.joint_state.position.empty())
+  {
+    ROS_ERROR_STREAM_NAMED("trajopt_planner", "Start state position is empty");
+    res.error_code.val = moveit_msgs::MoveItErrorCodes::INVALID_ROBOT_STATE;
+    return false;
+  }
+
+  if (req.start_state.joint_state.name.empty())
+  {
+    ROS_ERROR_STREAM_NAMED("trajopt_planner", "Start_state name is empty");
+    res.error_code.val = moveit_msgs::MoveItErrorCodes::INVALID_ROBOT_STATE;
+    return false;
+  }
+
+  if (not planning_scene->getRobotModel()->satisfiesPositionBounds(req.start_state.joint_state.position.data()))
```

Comment:
This check must use the `JointModelGroup`  specified in the `MotionPlanRequest` (`req.group_name`). Otherwise subgroups can't be used for planning.

Reply:
right. replaced it with the joint positions related to the joint_model_group